File: ProjectEuler/501-550/S549.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ProjectEuler
{
    public class S549 : ISolution
    {
        private static readonly int N = 100000000;

        public string GetAnswer()
        {
            var originPrimes = Utils.GenPrimes(N + 1);
            var primes = originPrimes.Where(l => l != 0).ToArray();

            var result = new long[N + 1];
            foreach (var p in primes)
            {
                result[p] = p;
            }

            for (int index = 0; primes[index] < N / 2; index++)
            {
                for (int j = 2; j < primes[index]; j++)
                {
                    var tmp = j * primes[index];
                    if (tmp > N)
                    {
                        break;
                    }
                    else
                    {
                        result[tmp] = primes[index];
                    }
                }
            }

            for (int index = 0; primes[index] < N / 6; index++)
            {
                for (int j = 2; j < primes[index] - 1; j++)
                {
                    long k = Math.Max(j + 1, primes[index] / j);
                    var test = j * k * primes[index];
                    if (test < 0 || test > N)
                    {
                        break;
                    }

                    for (; k < primes[index]; k++)
                    {
                        var tmp = j * k * primes[index];
                        if (tmp < 0 || tmp > N)
                        {
                            break;
                        }
                        else
                        {
                            result[tmp] = primes[index];
                        }
                    }

                }
            }

            for (int index = 0; primes[index] < N / 24; index++)
            {
                for (int j = 2; j < primes[index] - 2; j++)
                {
                    long k = Math.Max(j + 1, primes[index] / j);
                    var test = j * k * (k + 1) * primes[index];
                    if (test < 0 || test > N)
                    {
                        break;
                    }

                    for (; k < primes[index] - 1; k++)
                    {
                        long m = Math.Max(k + 1, primes[index] * primes[index] / (j * k));
                        var test2 = j * k * m * primes[index];
                        if (test2 < 0 || test2 > N)
                        {
                            break;
                        }

                        for (; m < primes[index]; m++)
                        {
                            var tmp = j * k * m * primes[index];
                            if (tmp < 0 || tmp > N)
                            {
                                break;
                            }
                            else
                            {
                                result[tmp] = primes[index];
                            }
                        }
                    }
                }
            }

            for (int i = 2; i <= N; i++)
            {
                if (result[i] == 0)
                {
                    var maps = Utils.Factorize(i, originPrimes);
                    long m = long.MinValue;
                    foreach (var map in maps)
                    {
                        var temp = GetMaxMByPrime(map);
                        if (temp > m)
                        {
                            m = temp;
                        }
                    }

                    result[i] = m;
                }
            }

            return result.Sum().ToString();
        }

        private static long GetMaxMByPrime(KeyValuePair<long, int> map)
        {
            long result = 0;
            long count = map.Value;
            while (count > 0)
            {
                result += map.Key;
                long cur = result;
                while (cur > 0)
                {
                    if (cur % map.Key == 0)
                    {
                        cur /= map.Key;
                        count--;
                    }
                    else
                    {
                        break;
                    }
                }
            }

            return result;
        }
    }
}

File: src/501-550/S549.cc
#include "S549.h"
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

namespace ProjectEuler
{
static int s_calc(int p, int k)
{
	int m = 0;
	int count = 0;
	while (count < k)
	{
		m += p;
		int temp = m;
		while (temp % p == 0)
		{
			count++;
			temp /= p;
		}
	}
	return m;
}

std::string S549::GetAnswer()
{
	int N = 100000000;
	std::vector<int> S(N + 1, 0);

	for (int i = 2; i <= N; ++i)
	{
		if (S[i] == 0)
		{
			// i is prime
			long long p = i;
			long long pk = p;
			int k = 1;

			while (pk <= N)
			{
				int val;
				if (k == 1)
				{
					val = p;
				}
				else
				{
					val = s_calc((int)p, k);
				}

				for (int j = (int)pk; j <= N; j += (int)pk)
				{
					if (S[j] < val)
					{
						S[j] = val;
					}
				}

				if (N / p < pk)
				{
					break;
				}
				pk *= p;
				k++;
			}
		}
	}

	long long sum = 0;
	for (int i = 2; i <= N; ++i)
	{
		sum += S[i];
	}

	return std::to_string(sum);
}
} // namespace ProjectEuler
